homeassistant/components/backblaze/backup.py

```diff
@@ -0,0 +1,228 @@
+    """Backblaze backup agent."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: BackblazeConfigEntry) -> None:
+        """Initialize the Backblaze backup sync agent."""
+        super().__init__()
+        self._bucket = entry.runtime_data.bucket
+        self._api = entry.runtime_data.api
+        self._hass = hass
+        self.name = entry.title
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file from Backblaze."""
+        if not await self.async_get_backup(backup_id):
+            raise BackupAgentError("Backup not found")
+
+        try:
+            downloaded_file = await self._hass.async_add_executor_job(
+                self._bucket.download_file_by_name, f"{backup_id}.tar"
+            )
+        except B2Error as err:
+            raise BackupAgentError(
+                f"Failed to download backup {backup_id}: {err}"
+            ) from err
+
+        if not downloaded_file.response.ok:
+            raise BackupAgentError(
+                f"Failed to download backup {backup_id}: HTTP {downloaded_file.response.status_code}"
+            )
+
+        # Use an executor to avoid blocking the event loop
+        for chunk in await self._hass.async_add_executor_job(
+            downloaded_file.response.iter_content, 1024
+        ):
+            yield chunk
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup."""
+
+        # Prepare file info metadata to store with the backup in Backblaze
+        file_info = {
+            "backup_id": backup.backup_id,
+            "database_included": str(backup.database_included).lower(),
+            "date": backup.date,
+            "extra_metadata": "###META###".join(
+                f"{key}{SEPARATOR}{val}" for key, val in backup.extra_metadata.items()
+            ),
+            "homeassistant_included": str(backup.homeassistant_included).lower(),
+            "homeassistant_version": backup.homeassistant_version,
+            "name": backup.name,
+            "protected": str(backup.protected).lower(),
+            "size": str(backup.size),
+        }
+        if backup.addons:
+            file_info["addons"] = "###ADDON###".join(
+                f"{addon.slug}{SEPARATOR}{addon.version}{SEPARATOR}{addon.name}"
+                for addon in backup.addons
+            )
+        if backup.folders:
+            file_info["folders"] = ",".join(folder.value for folder in backup.folders)
+
+        stream: AsyncIterator[bytes] = await open_stream()
+
+        try:
+            await self._hass.async_add_executor_job(
+                self._bucket.upload_bytes,
+                b"".join([chunk async for chunk in stream]),
```

Comment:
this means you're pushing the whole backup into memory. This won't work for bigger backups.

Does `upload_bytes` take an iterator? If so, you could write a sync iterator that uses `run_coroutine_threadsafe` to get a megabyte of data from the async iterator.

Reply:
Yeah that is the bad side here...

There is a method that takes any stream, which makes a lot of sense... (I expect that to be more common than an iterator for this)

But I have no clue on how to wrap an async iterator into a stream.

Working on this, made we wonder why we made this an iterator to begin with 😬

Reply:
I think I have an idea how to wrap it. Will whip something up during the day.

Reply:
I think I have an idea how to wrap it. Will whip something up during the day.